File: option_pricing.c
/* Adapted from the C++ version of the Monte Carlo method for pricing options found
at https://www.quantstart.com/articles/European-vanilla-option-pricing-with-C-via-Monte-Carlo-methods/ */

#include "option_pricing.h"

#define MAX(a, b) ((a) > (b) ? (a) : (b)) // Macro for max() fuunctionality 

// A simple implementation of the Box-Muller algorithm, used to generate
// gaussian random numbers - necessary for the Monte Carlo method below
double gaussian_box_muller(pcg32_random_t *rng) {
  double x = 0.0;
  double y = 0.0;
  double euclid_sq = 0.0;

  // Continue generating two uniform random variables
  // until the square of their "euclidean distance" 
  // is less than unity
  do {
    x = 2.0 * pcg32_random_r(rng) / (double)UINT32_MAX - 1;
    y = 2.0 * pcg32_random_r(rng) / (double)UINT32_MAX - 1;
    euclid_sq = x*x + y*y;
  } while (euclid_sq >= 1.0);

  return x*sqrt(-2*log(euclid_sq)/euclid_sq);
}

double monte_carlo_call_payoff_sum(pricing_params_t* params) {
  double S_adjust = params->S * exp(params->T*(params->r-0.5*params->v*params->v));
  double S_cur = 0.0;
  double payoff_sum = 0.0;

  for (int i=0; i<params->num_sims; i++) {
    double gauss_bm = gaussian_box_muller(&params->rng);
    S_cur = S_adjust * exp(sqrt(params->v*params->v*params->T)*gauss_bm);
    payoff_sum += MAX(S_cur - params->K, 0.0);
  }

  return payoff_sum;
}

double monte_carlo_put_payoff_sum(pricing_params_t* params) {
  double S_adjust = params->S * exp(params->T*(params->r-0.5*params->v*params->v));
  double S_cur = 0.0;
  double payoff_sum = 0.0;

  for (int i=0; i<params->num_sims; i++) {
    double gauss_bm = gaussian_box_muller(&params->rng);
    S_cur = S_adjust * exp(sqrt(params->v*params->v*params->T)*gauss_bm);
    payoff_sum += MAX(params->K - S_cur, 0.0);
  }

  return payoff_sum;
}

// wrapper function for the monte carlo call and put pricing functions
void* thread_pricing_wrapper(void* arg) {
  pricing_params_t* params = (pricing_params_t*)arg;
  params->call_payoff_sum = monte_carlo_call_payoff_sum(params);
  params->put_payoff_sum = monte_carlo_put_payoff_sum(params);
  return NULL;
}

int main(int argc, char* argv[]) {

  struct timespec start, end;
  clock_gettime(CLOCK_MONOTONIC, &start);

  int num_threads = (int)sysconf(_SC_NPROCESSORS_ONLN);
  uint64_t base_seed = (uint64_t)time(NULL);
  if (num_threads < 1) {
    num_threads = 1;
  }
  // First we create the parameter list                                                                               
  int num_sims = 10000000;   // Number of simulated asset paths
  num_sims = num_sims - (num_sims % num_threads); // remove any remainder from the number of simulations                                                   
  double S = 100.0;  // Option price                                                                                  
  double K = 100.0;  // Strike price                                                                                  
  double r = 0.05;   // Risk-free rate (5%)                                                                           
  double v = 0.2;    // Volatility of the underlying (20%)                                                            
  double T = 1.0;    // One year until expiry
  
  // each thread will calculate the call and put payoff sums for a portion of the simulations
  pthread_t threads[num_threads];
  pricing_params_t thread_params[num_threads];
  
  for (int i = 0; i < num_threads; i++) {
    thread_params[i] = (pricing_params_t){
      .num_sims = num_sims / num_threads, // portion of the simulations for each thread
      .S = S,
      .K = K,
      .r = r,
      .v = v,
      .T = T,
      .call_payoff_sum = 0.0,
      .put_payoff_sum = 0.0,
    };
    // Initialize PCG RNG with unique seed for each thread
    pcg32_srandom_r(&thread_params[i].rng, base_seed + (uint64_t)i, (uint64_t)i);
    pthread_create(&threads[i], NULL, thread_pricing_wrapper, (void*)&thread_params[i]);
  }

  for (int i = 0; i < num_threads; i++) {
    pthread_join(threads[i], NULL);
  }

  // Sum all payoff sums from all threads
  double total_call_payoff_sum = 0.0;
  double total_put_payoff_sum = 0.0;
  for (int i = 0; i < num_threads; i++) {
    total_call_payoff_sum += thread_params[i].call_payoff_sum;
    total_put_payoff_sum += thread_params[i].put_payoff_sum;
  }

  // Calculate final prices once from the aggregated payoff sums
  double call_price = (total_call_payoff_sum / (double)num_sims) * exp(-r*T);
  double put_price = (total_put_payoff_sum / (double)num_sims) * exp(-r*T);

  clock_gettime(CLOCK_MONOTONIC, &end);
  double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;

  // Finally we output the parameters and prices and other useful information
  printf("Elapsed time: %.6f seconds\n", elapsed);
  // printf("Number of threads: %d\n", num_threads);                                                                      
  // printf("Number of Paths: %d\n", num_sims);
  // printf("Underlying:      %.2f\n", S);
  // printf("Strike:          %.2f\n", K);
  // printf("Risk-Free Rate:  %.2f\n", r);
  // printf("Volatility:      %.2f\n", v);
  // printf("Maturity:        %.2f\n", T);

  printf("Call Price:      %.6f\n", call_price);
  printf("Put Price:       %.6f\n", put_price);

  return 0;
}

File: option_pricing.py
"""
Adapted from the C++ version of the Monte Carlo method for pricing options found
at https://www.quantstart.com/articles/European-vanilla-option-pricing-with-C-via-Monte-Carlo-methods/
"""

import random
import math
import time
import multiprocessing as mp

# A simple implementation of the Box-Muller algorithm, used to generate
# gaussian random numbers - necessary for the Monte Carlo method below.
def gaussian_box_muller(rng):
    x = 0.0
    y = 0.0
    euclid_sq = 0.0
    
    # Continue generating two uniform random variables
    # until the square of their "euclidean distance" 
    # is less than unity
    while True:
        x = 2.0 * rng.random() - 1
        y = 2.0 * rng.random() - 1
        euclid_sq = x * x + y * y
        if euclid_sq < 1.0:
            break
    
    return x * math.sqrt(-2 * math.log(euclid_sq) / euclid_sq)

# function to calculate the call option payoff sum using Monte Carlo simulation
def monte_carlo_call_payoff_sum(num_sims, S, K, r, v, T, seed):
    rng = random.Random(seed)
    
    S_adjust = S * math.exp(T * (r - 0.5 * v * v))
    payoff_sum = 0.0
    
    for i in range(num_sims):
        gauss_bm = gaussian_box_muller(rng)
        S_cur = S_adjust * math.exp(math.sqrt(v * v * T) * gauss_bm)
        payoff_sum += max(S_cur - K, 0.0)
    
    return payoff_sum

# function to calculate the put option payoff sum using Monte Carlo simulation
def monte_carlo_put_payoff_sum(num_sims, S, K, r, v, T, seed):
    rng = random.Random(seed)
    
    S_adjust = S * math.exp(T * (r - 0.5 * v * v))
    payoff_sum = 0.0
    
    for i in range(num_sims):
        gauss_bm = gaussian_box_muller(rng)
        S_cur = S_adjust * math.exp(math.sqrt(v * v * T) * gauss_bm)
        payoff_sum += max(K - S_cur, 0.0)
    
    return payoff_sum


def process_pricing_wrapper(args):
    num_sims, S, K, r, v, T, seed = args
    call_payoff_sum = monte_carlo_call_payoff_sum(num_sims, S, K, r, v, T, seed)
    put_payoff_sum = monte_carlo_put_payoff_sum(num_sims, S, K, r, v, T, seed)
    return call_payoff_sum, put_payoff_sum


def main():
    start_time = time.time()
    
    # Get number of CPU cores
    num_procs = mp.cpu_count()
    if num_procs < 1:
        num_procs = 1
    
    base_seed = int(time.time())
    
    # Parameters
    num_sims = 10000000  # Number of simulated asset paths
    num_sims = num_sims - (num_sims % num_procs)  # Remove remainder
    
    S = 100.0  # Option price (underlying)
    K = 100.0  # Strike price
    r = 0.05   # Risk-free rate (5%)
    v = 0.2    # Volatility of the underlying (20%)
    T = 1.0    # One year until expiry
    
    # Prepare arguments for each process
    sims_per_proc = num_sims // num_procs
    process_args = []
    for i in range(num_procs):
        seed = base_seed + i
        process_args.append((sims_per_proc, S, K, r, v, T, seed))
    
    # Run simulations in parallel
    with mp.Pool(processes=num_procs) as pool:
        results = pool.map(process_pricing_wrapper, process_args)
    
    # Sum all payoff sums from all processes
    total_call_payoff_sum = sum(result[0] for result in results)
    total_put_payoff_sum = sum(result[1] for result in results)
    
    # Calculate final prices once from the aggregated payoff sums
    call_price = (total_call_payoff_sum / num_sims) * math.exp(-r * T)
    put_price = (total_put_payoff_sum / num_sims) * math.exp(-r * T)
    
    elapsed = time.time() - start_time
    
    # Output results
    print("Elapsed time: {:.6f} seconds".format(elapsed))
    # print("Number of processes: {}".format(num_procs))
    # print("Number of Paths: {}".format(num_sims))
    # print("Underlying:      {:.2f}".format(S))
    # print("Strike:          {:.2f}".format(K))
    # print("Risk-Free Rate:  {:.2f}".format(r))
    # print("Volatility:      {:.2f}".format(v))
    # print("Maturity:        {:.2f}".format(T))
    print("Call Price:      {:.6f}".format(call_price))
    print("Put Price:       {:.6f}".format(put_price))


if __name__ == "__main__":
    main()

